### ISSUE TITLE
famers should be able to stop, resume or delete workloads running on their nodes

### ISSUE THREAD

**Issue by xmonader** [threefoldtech/home#1375]
That's a feature request from Kristof, specs still not there yet, but the current assumption it's about abuse, better option should be having the complaint filed through the DAO and the DAO should do the action

How can we identify the workload? if the node is shared, it can be problematic

**Comment by xmonader** [comment]
@LeeSmet @DylanVerstraete @muhamadazmy

**Comment by DylanVerstraete** [comment]
As long as we don't have Identity on chain I don't think there is a good solution for this. Their are problems on many levels, as you said:

- How do we identify which workload that needs to be stopped
- What prevents the user from re-creating the workload once it has stopped
- There is no way to punish bad behaviour atm becuase there is no reputation system
- ...

Before we adress the above issues I think it's irrelevant to start thinking about stopping a workload in my opinion.

**Comment by scottyeager** [comment]
Complaints related to public IPs can be linked to the workload that reserves the public IP without much trouble. I don't see that gateway workloads currently have their subdomain logged on TF Chain, but this seems relatively simple to achieve. The really difficult kind of traffic to link to a workload, I think, is outbound traffic (node used in DDoS attack, for example).

Regarding the identity concerns, the simplest step I see is requiring the user to lock up some TFT in order to activate their twin and be eligible to create deployments. For example:

* User deposits 2000 TFT and 1000 of them are locked for account activation
* User creates some deployments and spends their liquid 1000 TFT
* The deployments remain active and drain from the locked balance, but no new deployments can be created
* With 500 locked TFT left, user deletes all contracts from their twin
* User can now initiate unlocking and two weeks later their remaining 500 TFT can be transferred

While not a perfect solution, this creates some transient cost for creating accounts that would help with concerns about "sybil attacks" against farmers who are trying to regulate the deployments on their nodes.

**Comment by DylanVerstraete** [comment]
I don't believe depositing this amount of TFT is a great solution, it will only make the entry level harder for normal ppl. 

As said in the above comment, there should be some KYC which links a user's twin to a physical person. If we do that, we can create a reputation system and allow the DAO to blacklist users. This whole identity concept can be even expanded with linking users to a cellphone for example, so they cannot create more than one twin.

**Comment by scottyeager** [comment]
Sure, we can discuss how to achieve Sybil resistance elsewhere, but Dylan's original point about this is still valid. Without a compatible plan to preventing the deployment from popping up again under a different account, this feature won't have any teeth, so including that as a consideration is essential.

As for defining the kinds of abuses that a farmer needs to be empowered to respond to, the most obvious scope is whatever is both broadly illegal and can potentially generate some enforcement action towards the farmer. Beyond illegal content, participation of the farmers nodes in DDoS attacks is something that authorities take [rather seriously](https://www.nationalcrimeagency.gov.uk/?view=article&id=243:ddos-attacks-are-illegal&catid=2) and can't be easily pinned to a public IP reservation. These categories come to mind:

1. Accessing, hosting, and sharing content that is illegal to posses
2. Distributing copyrighted material
3. DDoS attacks and other spam
4. Acting as a proxy for hacking/cracking other systems

I wonder to what extent the already logged network consumption data could be helpful in identifying problematic workloads, since all traffic over the public internet is now logged for billing by Zos. Correlating traffic spikes for a workload with the timing of attacks could provide a useful, if not foolproof way to figure out which workload is responsible. 

Modifying Zos to provide more data, of a generalized sort so as not to be a privacy concern for legitimate users, is another step we could consider. A simple count of the number of network connections opened by a workload, on a relatively rough grained time interval, could also be a helpful tool for this purpose.

**Comment by LeeSmet** [comment]
Outgoing connections are already impossible to verify. In fact, as a malicious actor, it is possible to make my traffic look completely indistinguishable from another actor on the same node. This is the result of a combination of virtual hosting / CDN's and the advance of privacy protective measures implemented in TLS1.3/DNS over https. In combination with metrics.grid.tf, you can make your own traffic completely blend in with existing traffic. This is of course specific for accessing content over a web browser, but this will likely be a larger share of the "abuse" in this sense.

Other than that, if we start to do more in depth traffic analysis, then we are essentially becoming big brother ourselves, which is very much not what we want. Even if we did this, it should also be noted that these kind of measures are easily defeated by an active attacker, which will always be the case anyhow.

The one way to start identifying workloads for outgoing connections would be to actively track the masquerading in  the NDMZ, such that we can keep track of the (src ip, src port, dst ip, dst port) quads for all connections. src ip in this context will be the carrier grade NAT ip assigned in the network resource. This is hidden from the user, but unique on the node (the user chosen IP in private networks can conflict). This could work technically, but it also approaches techniques used by governments to identify citizens who don't abide by their ideas, i.e. this is a very big nono from a privacy perspective. Also, active attackers will probably intermix garbage traffic to legitimate sources, so keeping track of this for long enough for all connections is very quickly going to result in a very huge dataset. 

The previous all doesn't actually matter by virtue of the first paragraph, as all we have is a destination IP, which could host multiple sites, and therefore we don't know what content the workload is actually going towards. If the attacker choses a node with existing workloads, these just might have connections to this same IP (especially in case CDN's are in play), accessing completely legit content.
